core/src/connection.rs

```diff
@@ -193,12 +198,47 @@ impl ConnectedPoint {
     /// For `Dialer`, this modifies `address`. For `Listener`, this modifies `send_back_addr`.
     pub fn set_remote_address(&mut self, new_address: Multiaddr) {
         match self {
-            ConnectedPoint::Dialer { address } => *address = new_address,
+            ConnectedPoint::Dialer { address, .. } => *address = new_address,
             ConnectedPoint::Listener { send_back_addr, .. } => *send_back_addr = new_address,
         }
     }
 }
 
+/// Whether a connection is dialed _as a listener_.
+///
+/// This option is needed for NAT and firewall hole punching.
+///
+/// The concrete realization of this option depends on the transport
+/// protocol. E.g. in the case of TCP, both endpoints dial each other,
+/// resulting in a _simultaneous open_ TCP connection. On this new
+/// connection both endpoints assume to be the dialer of the connection.
+/// This is problematic during the connection upgrade process where an
+/// upgrade assumes one side to be the listener. With the help of this
+/// option, both peers can negotiate the roles (dialer and listener) for
+/// the new connection ahead of time, through some external channel, and
+/// thus have one peer dial the other as a dialer and one peer dial the
+/// other _as a listener_.
+#[derive(Debug, Default, Copy, Clone, Hash, Eq, PartialEq)]
+pub struct DialAsListener(bool);
```

Comment:
I introduced this newtype instead of passing a `bool` through the various layers.

I like the fact that it (a) increases type safety and (b) carries its documentation. Unfortunately it is not particularly intuitive, nor idiomatic (never seen a newtype around a `bool`) nor convenient (e.g. see `if as_listener == false`).

@thomaseizinger you usually have good ideas when it comes to idiomatic Rust. Any alternatives that come to your mind?

Reply:
What do you think of:

```rust
    Dialer {
        /// An optional override if the desired role for the upgrade process has already been negotiated out of band.
		////
		/// If this is set, the endpoint resulting from a successful dial should assume the given role instead of inferring it from the underlying transport.
        role_override: Option<Endpoint>,
    },
```

We could be even more descriptive and do:

```rust
enum UpgradeRole {
	InferredFromTransport,
	NegotiatedOutOfBand {
		role: Endpoint
	}
}
// -------------
Dialer {
    /// The strategy, how the role within the upgrade process should be determined.
    upgrade_role: UpgradeRole,
},
```

Reply:
```rust
    Dialer {
        /// An optional override if the desired role for the upgrade process has already been negotiated out of band.
		////
		/// If this is set, the endpoint resulting from a successful dial should assume the given role instead of inferring it from the underlying transport.
        role_override: Option<Endpoint>,
    },
```

This is a good idea, especially as it does not introduce a new type. Though it does allow for duplication of information and thus potentially confusion? `PendingPoint::Dialer { role_override: Some(Endpoint::Dialer) }` is the same as `PendingPoint::Dialer { role_override: None }`.

Reply:
```rust
enum UpgradeRole {
	InferredFromTransport,
	NegotiatedOutOfBand {
		role: Endpoint
	}
}
```

Note that this whole out-of-band role negotiation does not only concern upgrades, but some transport protocols as well. The TCP transport implementation does not require this information as it will open a simultaneous open TCP connection either way. But QUIC does. In the case of QUIC the listener sends a random-garbage packet instead of a connection-initiation packet.

Reply:
Back to your suggestion of `role_override`, how would the signatures of `Network::dial` and `Peer::dial` look like?

```
    pub fn dial(
        &mut self,
        address: &Multiaddr,
        handler: THandler,
        role_override: Option<Endpoint>,
    ) -> Result<ConnectionId, DialError<THandler>>
```

I find this descriptive but again confusing due to the duplication of information where `Some(Endpoint::Dialer)` is the same as `None`.